### ISSUE TITLE
Fix empty allocations

### ISSUE THREAD

**Issue by fmoessbauer** [dash-project/dash#239]
Currently GlobMem allocations hang if at least one unit does not contribute any local elements. This is due to collective MPI operations in DART which are skipped for units without local elements. Furthermore the implementation of memory allocation does not comply to the specifications made in the header.

For details also have a look at my comment [here](https://github.com/dash-project/dash/issues/226#issuecomment-273561852) in issue #226 .

Therefor I recommend to fix the implementation and clarify the specification as follows:

- all allocators have to support empty allocations
- all units of a team take part in a team-wide allocation
- all allocation calls have to be passed to the down-most (DART) allocator
- a dash allocator must not lessen the synchronisation requirements of the down-most allocator

**Comment by devreal** [comment]
I just checked the DART code for skipped collective operations. I cannot see that DART itself skips collectives on empty allocations. So it's just that DASH is skipping collective DART calls on empty allocations?

**Comment by fmoessbauer** [comment]
There are two sides of the problem:
1. DASH skips the collective operation `dart_team_memalloc_aligned`.
2. DART does not allow empty allocations, as `MPI_win_attach` does not work for empty windows (not sure if this is the bug, but it's the error MPI reports). However a simple `if(nelem == 0){++nelem;}` in DART fixes the problem. But this is definitly not the right way to solve the problem.

**Comment by devreal** [comment]
Can you please paste the MPI error message? At least the standard does not seem to enforce a size greater zero for `MPI_Win_attach`.

**Comment by fmoessbauer** [comment]
Executed using the openmpi (v1.10.5) docker container:

```
$>mpirun -n 4 --allow-run-as-root bin/dash-test-mpi
...
[  RUN     ] FindTest.EmptyContainer
...
[cffe0d848854:11978] *** An error occurred in MPI_Win_attach
[cffe0d848854:11978] *** reported by process [1494548481,2]
[cffe0d848854:11978] *** on win pt2pt window 7
[cffe0d848854:11978] *** MPI_ERR_ARG: invalid argument of some other kind
[cffe0d848854:11978] *** MPI_ERRORS_ARE_FATAL (processes in this win will now abort,
[cffe0d848854:11978] ***    and potentially your MPI job)
```

See branch [bug-239-allocate](https://github.com/dash-project/dash/compare/bug-239-allocate) for my implementation.

**Comment by devreal** [comment]
Please see the PR and check whether all cases are addressed (I'm not too familiar with the DASH side of global allocations but at least it fixes the test in question).

**Comment by fmoessbauer** [comment]
Looks good to me. Thanks for the fix.

**Comment by fmoessbauer** [comment]
After reading #221 I have some concerns that the way we fixed this does not comply with the assumptions of `dash::CollectiveAllocator` and hence `dart_team_memalloc_aligned()`. The documentation is not clear at this point, whether all units have to call the allocator with an equal number of local elements. According to @fuerlinger's comment, this is the case.

However if that is the case, the `GlobMem` has to be allocated for `pattern.local_capacity()` elements and `matrix.lend()` has to point at the last "usable" element in the memory (not the last allocated). However this is not really useful, as using pointer arithmetics on the GlobPtr is still not possible as the global memory space contains holes. Furthermore the memory usage can drop to at least 50%.

**Comment by fuerlinger** [comment]
As mentioned, the original idea was that `dart_team_memalloc_aligned()` is used with requesting an identical amount of memory on each unit and I suggest that we keep it that way. We can, however, add a version that allows for a different amount on each unit and we could call that version `dart_team_memalloc()`. This might be useful for cases where the amount of memory varies a lot between units and where a user is either willing to have global indexing be slower or forgo it entirely and only wishes to do local indexing (locally and remotely). A data structure that implements this idea was tentatively named a "directory" (might not be a great name). Schematically:

![snip](https://cloud.githubusercontent.com/assets/475258/22243682/a79badf8-e228-11e6-989e-02d46a102ce8.PNG)

**Comment by rkowalewski** [comment]
Maybe there is a chance to rename the aligned with a better term, e.g. `dart_team_memalloc_symmetric`. Aligned is misnamed here IMHO.

Regarding the issues as described by @fmoessbauer:

- A simple idea is to add some trait informtation to `GlobPtr` (e.g. `GlobPtr::memory_traits::symmetric`).  Simple pointer arithmetic should be well-defined if this trait is satisfied, otherwise undefined. It sounds maybe too simple but we can figure it out in a dedicated discussion.
- Regarding the memory usage I would not be too concerned. Because if we allocate memory it does not mean that the memory is actually mapped. Only if we touch it. As long as we care about this memory usage should be fine.

In the case of non-symmetric memory allocation the `lend`pointer should point to the usable memory as described above.

**Comment by fuchsto** [comment]
@fuerlinger How about `dart_team_coalloc`? It's shorter and that's what it does.

Now some lengthy Wisenheimer remark:
Traits are not part of the type they describe, they are templates specialized for types.
For example:

~~~c++
template <typename Iterator>
void duckduck(Iterator it) {
  typedef typename std::iterator_traits<Iterator>::value_type  referenced_value_t;
  // ...
}
~~~

I strongly oppose the notion of *memory_traits* for iterators.
So in your example, it's something like:

~~~c++
constexpr bool is_iterator_on_symmetric_mem = dash::memory_traits<GlobIterType>::symmetric::value
~~~

But actually again, symmetric allocation it's not a property of the iterator, it's defined by the allocator of its memory space. So we need something in the line of:

~~~c++
typedef typename dash::iterator_traits<GlobIterType>::allocator_type AllocType;
typedef typename dash::allocator_traits<AllocType>::memory_type MemType;
// and finally:
constexpr bool is_sym_memory = tynename dash::memory_traits<MemType>::is_symmetric::value;
~~~

There is no equivalent in the STL: iterators don't know about allocation and memory because that's the container's (or its allocator's) concern, and STL iterators do not allow to resolve the type of their container.

Long story short, I suggest not to describe the *cause* of the iterator type's semantic restrictions, but the semantic restrictions themselves.

So rather than describing
"the iterator references a memory space that has not been aquired from symmetric allocation so there is no co-index behind it"
as `memory_traits<GlobIter>` implies, I would state
"this iterator is not co-indexed".

~~~c++
dash::iterator_traits<GlobIterTytpe>::is_coindex
~~~

**Comment by fmoessbauer** [comment]
As the initial bug is fixed, I close this issue. Maybe we should move the discussion to a dedicated issue.
